File: python/src/patterns/creational/singleton/thread_safe_singleton.py
import threading


class ThreadSafeSingleton:
    _instance = None
    _lock = threading.Lock()

    def __init__(self) -> None:
        # prevent multiple instantiation by raising an exception
        if ThreadSafeSingleton._instance is not None:
            raise Exception("Cannot instantiate directly.")
        # init logic here

    @staticmethod
    def get_instance() -> "ThreadSafeSingleton":
        # only create once (the first time this method is called)
        # thread lock to allow one thread at a time
        with ThreadSafeSingleton._lock:
            if ThreadSafeSingleton._instance is None:
                ThreadSafeSingleton._instance = ThreadSafeSingleton()
        return ThreadSafeSingleton._instance

File: java/src/patterns/creational/singleton/ThreadSafeSingleton.java
package patterns.creational.singleton;

class ThreadSafeSingleton {
    // single static instance
    private static ThreadSafeSingleton instance;

    // private constructor prevents object instantiation
    private ThreadSafeSingleton() {
        // init logic here
    }

    // synchronized keyword to allow one thread at a time
    public static synchronized ThreadSafeSingleton getInstance() {
        // only create once (the first time this method is called)
        if (instance == null) {
            instance = new ThreadSafeSingleton();
        }
        return instance;
    }
}
